tests/rptest/tests/cluster_linking_test_base.py

```diff
@@ -44,6 +47,184 @@
+
+            if len(source) != len(target):
+                return None
+            for source_partition, target_partition in zip(source, target):
+                if source_partition.id != target_partition.id:
+                    return None
+                lag = source_partition.high_watermark - target_partition.high_watermark
+                self.logger.debug(
+                    f"Partition {topic}/{source_partition.id} - source: ({source_partition}), target: ({target_partition}) lag: {lag}"
+                )
+                ret[topic][source_partition.id] = lag
+
+        return ret
+
+    def total_lag(self, lag_dict) -> float:
+        if lag_dict is None:
+            return float("inf")
+        total = 0
+        for part_lags in lag_dict.values():
+            for lag in part_lags.values():
+                total += lag
+        return total
+
+    def replication_progress_validator(self, progress_timeout=60, backoff_delay=5):
+        total_last_lag = float("inf")
+        last_progress_time = time.time()
+
+        while True:
+            # Check replication progress
+            current_lag = self._calculate_partition_lag()
+            total_current_lag = self.total_lag(current_lag)
+            now = time.time()
+            if total_current_lag < total_last_lag or total_current_lag == 0:
+                self.logger.debug(
+                    f"Replication making progress - current_lag: {total_current_lag}, last_lag: {total_last_lag}"
+                )
+                last_progress_time = now
+
+            total_last_lag = total_current_lag
+            if time.time() - last_progress_time > progress_timeout:
+                self.logger.error(
+                    f"No replication progress for {progress_timeout}s, last lag: {total_last_lag}, current lag: {total_current_lag}"
+                )
+                raise Exception("Replication stalled")
+
+            if self.producer_finished and total_current_lag == 0:
+                self.logger.info(
+                    f"Replication finished, it took {time.time() - self.producer_finished_ts}s to catch up after producer finished"
+                )
+                break
+
+            time.sleep(backoff_delay)
+
+    def consumer_groups_state_consistent(self):
+        source_groups = self.source_rpk.group_list()
+        self.logger.debug(f"Source consumer groups: {source_groups}")
+        for g in source_groups:
+            source_d = self.source_rpk.group_describe(g.group)
+            target_d = self.target_rpk.group_describe(g.group)
+            target_partitions = {(p.topic, p.partition): p for p in target_d.partitions}
+
+            for p in source_d.partitions:
+                key = (p.topic, p.partition)
+                if key not in target_partitions:
+                    self.logger.error(
+                        f"Group {g.group} partition {key} offset commit not found in target"
+                    )
+                    return False
```

Comment:
nitpick: might be nice to accumulate errors rather than short circuit the loop, so we get a full picture of the group on failure.